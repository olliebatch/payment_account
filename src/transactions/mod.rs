mod read;

pub use read::*;
use rust_decimal::Decimal;

#[derive(serde::Serialize, serde::Deserialize, Clone, Copy, Debug, Eq, PartialEq, Hash)]
#[serde(rename_all = "lowercase")]
pub enum TransactionType {
    Deposit,
    Withdrawal,
    Dispute,
    Resolve,
    Chargeback,
}

#[derive(serde::Deserialize, serde::Serialize, Copy, Debug, PartialEq, Clone)]
pub struct Transaction {
    pub payment_type: TransactionType,
    pub client: u16,
    pub tx: u32,
    pub amount: Option<Decimal>,
}

impl Transaction {
    pub fn round(self) -> Self {
        if let Some(amount) = self.amount {
            let rounded = amount.round_dp(4);
            Transaction {
                amount: Some(rounded),
                ..self
            }
        } else {
            self
        }
    }

    #[cfg(test)]
    pub fn new_deposit(client: u16, tx: u32, amount: Option<Decimal>) -> Self {
        Transaction {
            payment_type: TransactionType::Deposit,
            client,
            tx,
            amount,
        }
    }
    #[cfg(test)]
    pub fn new_withdrawal(client: u16, tx: u32, amount: Option<Decimal>) -> Self {
        Transaction {
            payment_type: TransactionType::Withdrawal,
            client,
            tx,
            amount,
        }
    }
    #[cfg(test)]
    pub fn new_dispute(client: u16, tx: u32) -> Self {
        Transaction {
            payment_type: TransactionType::Dispute,
            client,
            tx,
            amount: None,
        }
    }
    #[cfg(test)]
    pub fn new_resolve(client: u16, tx: u32) -> Self {
        Transaction {
            payment_type: TransactionType::Resolve,
            client,
            tx,
            amount: None,
        }
    }
    #[cfg(test)]
    pub fn new_chargeback(client: u16, tx: u32) -> Self {
        Transaction {
            payment_type: TransactionType::Chargeback,
            client,
            tx,
            amount: None,
        }
    }
}
