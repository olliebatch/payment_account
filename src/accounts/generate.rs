use crate::accounts::ClientAccount;
use crate::transactions::{Transaction, TransactionType};
use std::collections::HashMap;

pub fn generate_client_accounts(transactions: Vec<Transaction>) -> HashMap<u16, ClientAccount> {
    let mut client_accounts: HashMap<u16, ClientAccount> = HashMap::new();

    for transaction in transactions {
        let account = client_accounts
            .entry(transaction.client)
            .or_insert(ClientAccount::new(transaction.client));

        let client = transaction.client;

        if transaction.payment_type == TransactionType::Deposit {
            let updated_account = account.clone().deposit_to_account(transaction);
            client_accounts.insert(client, updated_account);
        } else if transaction.payment_type == TransactionType::Withdrawal {
            let updated_account = account.clone().withdraw_from_account(transaction);
            client_accounts.insert(client, updated_account);
        }
    }
    client_accounts
}
