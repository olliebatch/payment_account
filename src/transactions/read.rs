use std::fs::File;

use crate::config::Config;
use crate::transactions::Transaction;
use crate::BoxError;

pub fn read_transaction_file(config: &Config) -> Result<Vec<Transaction>, BoxError> {
    let file = File::open(&config.transaction_file)?;
    let mut rdr = csv::Reader::from_reader(file);

    let mut transactions: Vec<Transaction> = vec![];
    for result in rdr.records() {
        let record = result?;
        let transaction: Transaction = record.deserialize(None)?;
        transactions.push(transaction);
    }

    Ok(transactions)
}
