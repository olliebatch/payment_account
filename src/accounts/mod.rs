use crate::transactions::Transaction;

pub struct ClientAccount {
    client: u16,
    available: f32,
    held: f32,
    total: f32,
    locked: bool,
}

impl ClientAccount {
    pub fn new(client_id: u16) -> Self {
        // start a new account with 0 balance and unlocked
        ClientAccount {
            client: client_id,
            available: 0.0,
            held: 0.0,
            total: 0.0,
            locked: false,
        }
    }

    #[cfg(test)]
    pub fn new_for_testing(client_id: u16, balance: f32) -> Self {
        ClientAccount {
            client: client_id,
            available: balance,
            held: 0.0,
            total: balance,
            locked: false,
        }
    }

    pub fn deposit_to_account(self, transaction: Transaction) -> Self {
        let new_avail = self.available + transaction.amount;
        let total = self.total + transaction.amount;
        ClientAccount {
            available: new_avail,
            total,
            ..self
        }
    }
    pub fn withdraw_from_account(self, transaction: Transaction) -> Self {
        let new_avail = self.available - transaction.amount;
        let total = self.total - transaction.amount;
        ClientAccount {
            available: new_avail,
            total,
            ..self
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::accounts::ClientAccount;
    use crate::transactions::Transaction;
    use rstest::rstest;

    #[rstest()]
    fn add_deposit_test() {
        let client_account = ClientAccount::new(1);
        let transaction = Transaction::new_deposit(1, 1, 1.5);

        let updated = client_account.deposit_to_account(transaction);
        assert_eq!(updated.available, 1.5);
        assert_eq!(updated.total, 1.5)
    }

    #[rstest()]
    fn withdraw_funds_test() {
        let client_account = ClientAccount::new_for_testing(1, 5.0);
        let transaction = Transaction::new_withdrawal(1, 1, 1.5);

        let updated = client_account.withdraw_from_account(transaction);
        assert_eq!(updated.available, 3.5);
        assert_eq!(updated.total, 3.5)
    }
}
