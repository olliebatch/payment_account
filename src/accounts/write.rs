use crate::accounts::{ClientAccount, ExternalClientAccount};
use crate::BoxError;
use csv::Writer;
use std::collections::HashMap;
use std::io;

pub fn write_to_stdout(accounts: HashMap<u16, ClientAccount>) -> Result<(), BoxError> {
    let mut wtr = Writer::from_writer(io::stdout());

    for account in accounts.values() {
        let external_account: ExternalClientAccount = account.into();
        wtr.serialize(external_account)?;
    }

    wtr.flush()?;
    Ok(())
}
