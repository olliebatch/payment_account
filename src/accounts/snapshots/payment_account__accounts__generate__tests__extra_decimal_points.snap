---
source: src/accounts/generate.rs
expression: "serde_json::json!(result)"

---
{
  "1": {
    "available": 1.1235,
    "client": 1,
    "held": 0.0,
    "locked": false,
    "total": 1.1235,
    "transactions": [
      {
        "amount": 1.1235,
        "client": 1,
        "payment_type": "deposit",
        "tx": 1
      },
      {
        "amount": 2.1235,
        "client": 1,
        "payment_type": "deposit",
        "tx": 3
      },
      {
        "amount": 2.1235,
        "client": 1,
        "payment_type": "withdrawal",
        "tx": 4
      }
    ]
  },
  "2": {
    "available": 0.0,
    "client": 2,
    "held": 0.0,
    "locked": false,
    "total": 0.0,
    "transactions": [
      {
        "amount": 2.1235,
        "client": 2,
        "payment_type": "deposit",
        "tx": 2
      },
      {
        "amount": 2.1235,
        "client": 2,
        "payment_type": "withdrawal",
        "tx": 5
      }
    ]
  }
}
